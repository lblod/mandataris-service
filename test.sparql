  PREFIX mandaat: <http://data.vlaanderen.be/ns/mandaat#>
  PREFIX mu: <http://mu.semte.ch/vocabularies/core/>
  PREFIX besluit: <http://data.vlaanderen.be/ns/besluit#>  
  PREFIX person: <http://www.w3.org/ns/person#>
  PREFIX mu: <http://mu.semte.ch/vocabularies/core/>
  PREFIX person: <http://www.w3.org/ns/person#>
  PREFIX persoon: <http://data.vlaanderen.be/ns/persoon#>
  PREFIX foaf: <http://xmlns.com/foaf/0.1/>
  PREFIX adms: <http://www.w3.org/ns/adms#>
  PREFIX skos: <http://www.w3.org/2004/02/skos/core#>

   INSERT DATA {
    GRAPH <http://mu.semte.ch/graphs/besluiten-consumed974816591f269bb7d74aa1720922651529f3d3b2a787f5c60b73e5a0384950a4/LoketLB-mandaatGebruiker> {
      <http://data.lblod.info/id/personen/1234567689> a person:Person;
          mu:uuid """1234567689""";
          persoon:gebruikteVoornaam """Jonas""";
          foaf:familyName """De tester""";
          persoon:geslacht <http://publications.europa.eu/resource/authority/human-sex/FEMALE>;
          foaf:name """JJ""";
          persoon:heeftGeboorte <http://data.lblod.info/id/geboortes/d9e8f8fa-8f5e-4d97-920e-70cff639de11>.
    }
  }

