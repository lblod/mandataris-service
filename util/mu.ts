import {
  sparqlEscapeUri as mu_sparqlEscapeUri,
  sparqlEscapeString as mu_sparqlEscapeString,
  sparqlEscapeDateTime as mu_sparqlEscapeDateTime,
} from 'mu';

export function sparqlEscapeString(stringValue: string) {
  if (!stringValue || typeof stringValue !== 'string') {
    throw Error(
      `Could not sparql escape string from passed value: ${stringValue}`,
    );
  }

  return mu_sparqlEscapeString(stringValue);
}

export function sparqlEscapeUri(uri: string) {
  if (!uri || typeof uri !== 'string') {
    throw Error(`Could not sparql escape uri from passed value: ${uri}`);
  }

  return mu_sparqlEscapeUri(uri);
}

export function sparqlEscapeDateTime(date: Date) {
  if (!date) {
    throw Error(`Could not sparql escape datetime from passed value: ${date}`);
  }

  return mu_sparqlEscapeDateTime(date);
}
