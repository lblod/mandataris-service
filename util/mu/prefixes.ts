import { sparqlEscapeUri } from 'mu';

const mapping = {
  mu: 'http://mu.semte.ch/vocabularies/core/',
  besluit: 'http://data.vlaanderen.be/ns/besluit#',
  mandaat: 'http://data.vlaanderen.be/ns/mandaat#',
  org: 'http://www.w3.org/ns/org#',
  lmb: 'http://lblod.data.gift/vocabularies/lmb/',
  ext: 'http://mu.semte.ch/vocabularies/ext/',
  person: 'http://www.w3.org/ns/person#',
  extlmb: 'http://mu.semte.ch/vocabularies/ext/lmb/',
  xsd: 'http://www.w3.org/2001/XMLSchema#',
  astreams: 'http://www.w3.org/ns/activitystreams#',
  skos: 'http://www.w3.org/2004/02/skos/core#',
  persoon: 'http://data.vlaanderen.be/ns/persoon#',
  adms: 'http://www.w3.org/ns/adms#',
  regorg: 'https://www.w3.org/ns/regorg#',
  muAccount: 'http://mu.semte.ch/vocabularies/account/',
  foaf: 'http://xmlns.com/foaf/0.1/',
  dcterms: 'http://purl.org/dc/terms/',
  dct: 'http://purl.org/dc/terms/',
  nfo: 'http://www.semanticdesktop.org/ontologies/2007/03/22/nfo#',
  dbp: 'http://dbpedia.org/ontology/',
  nie: 'http://www.semanticdesktop.org/ontologies/2007/01/19/nie#',
  session: 'http://mu.semte.ch/vocabularies/session/',
  generiek: 'http://data.vlaanderen.be/ns/generiek#',
  rdf: 'http://www.w3.org/1999/02/22-rdf-syntax-ns#',
  owl: 'http://www.w3.org/2002/07/owl#',
  schema: 'http://schema.org/',
};

export function getPrefixesForQuery(sparqlQuery: string) {
  const prefixes: string[] = [];

  for (const key in mapping) {
    if (sparqlQuery.includes(`${key}:`)) {
      prefixes.push(`PREFIX ${key}: ${sparqlEscapeUri(mapping[key])}`);
    }
  }

  return prefixes;
}
