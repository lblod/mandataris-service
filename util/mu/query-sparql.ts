import { query as mu_query, update as mu_update } from 'mu';
import {
  querySudo as mu_querySudo,
  updateSudo as mu_updateSudo,
} from '@lblod/mu-auth-sudo';

import { getPrefixesForQuery } from './prefixes';

export async function query(sparqlQuery: string) {
  return await mu_query(getQueryWithPrefixes(sparqlQuery));
}

export async function update(sparqlQuery: string) {
  return await mu_update(getQueryWithPrefixes(sparqlQuery));
}
export async function querySudo(sparqlQuery: string) {
  return await mu_querySudo(getQueryWithPrefixes(sparqlQuery));
}

export async function updateSudo(sparqlQuery: string) {
  return await mu_updateSudo(getQueryWithPrefixes(sparqlQuery));
}

function getQueryWithPrefixes(sparqlQuery) {
  const startTime = performance.now();
  const prefixes = getPrefixesForQuery(sparqlQuery);
  console.log(
    `\n\n\tPERFORMANCE: getting the prefixes took: ${
      performance.now() - startTime
    } ms \n\n`,
  );
  return `
    # generated prefixes
    ${prefixes.join('\n')}
    # end generated prefixes
    ${sparqlQuery}
  `;
}
