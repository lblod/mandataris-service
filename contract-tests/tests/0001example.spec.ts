import 'dotenv/config';

import { describe, expect, test, beforeAll } from '@jest/globals';
import {
  mockLogin,
  userRequest,
  clearDeltas,
  expectQueryToHaveNoResults,
  getDeltas,
  runSudoQuery,
} from 'contract-tests';
import { beforeEach } from 'node:test';
import { sparqlEscapeUri } from 'mu';

describe('fractions', () => {
  beforeAll(async () => {
    await mockLogin(
      'http://data.lblod.info/id/bestuurseenheden/5116efa8-e96e-46a2-aba6-c077e9056a96',
      'http://data.lblod.info/id/accounts/1234',
      'LoketLB-mandaatGebruiker',
    );
  });

  beforeEach(async () => {
    await clearDeltas();
  });

  test('get mandataris fraction', async () => {
    const { body } = await userRequest(
      'GET',
      'http://target/mandatarissen/231e42a4-25a6-424b-9885-c16bea74a545/fracties',
    );
    expect(body).toMatchSnapshot();
    expect(await getDeltas()).toHaveLength(0);
  });

  test('update mandataris publication status', async () => {
    const mandatarisUris = [
      'http://data.lblod.info/id/mandatarissen/231e42a4-25a6-424b-9885-c16bea74a545',
      'http://data.lblod.info/id/mandatarissen/2f40df94-a8c6-47bc-8771-493c6b5f29bc',
    ];
    const bekrachtigd =
      'http://data.lblod.info/id/concept/MandatarisPublicationStatusCode/9d8fd14d-95d0-4f5e-b3a5-a56a126227b6';
    // don't have mu but as it's not application code we can be a bit more lenient
    const query = `
      PREFIX lmb: <http://lblod.data.gift/vocabularies/lmb/>
      PREFIX ext: <http://mu.semte.ch/vocabularies/ext/>

      SELECT * WHERE {
        VALUES ?mandataris {
          ${mandatarisUris.map((uri) => sparqlEscapeUri(uri)).join('\n')}
        }
        GRAPH ?g {
          ?mandataris lmb:hasPublicationStatus ${sparqlEscapeUri(bekrachtigd)} .
        }
        ?g ext:ownedBy ?someone.
      } ORDER by ?mandataris
    `;
    await expectQueryToHaveNoResults(query);
    const { body } = await userRequest(
      'POST',
      'http://target/mandatarissen/bulk-set-publication-status',
      {
        decision: 'http://data.lblod.info/id/besluit/1',
        statusUri: bekrachtigd,
        mandatarissen: mandatarisUris.map((uri) => uri.split('/').pop()),
      },
    );
    expect(body).toMatchSnapshot();
    const result = await runSudoQuery(query);
    expect(result).toMatchSnapshot();
    expect(await getDeltas()).toMatchSnapshot();
  });

  test('update-mandataris-fractie', async () => {
    // change someone's fraction before
    await runSudoQuery(`
      PREFIX org: <http://www.w3.org/ns/org#>
      DELETE {
        GRAPH ?g {
          ?membership org:organization ?old .
        }
      }
      INSERT {
        GRAPH ?g {
          ?membership org:organization <http://data.lblod.info/id/fracties/673C4FDF9AA6EE1DF1856B4E> .
        }
      }
      WHERE {
        VALUES ?membership {
          <http://data.lblod.info/id/lidmaatschappen/391c5846-659e-4742-86b6-b1b6f30883ff>
        }
        GRAPH ?g {
          ?membership org:organization ?old .
        }
      }`);

    await clearDeltas();
    const { body } = await userRequest(
      'PUT',
      'http://target/fracties/5F914115CE511800080003A4/current-fractie',
    );
    expect(body).toMatchSnapshot();
    expect(await getDeltas()).toMatchSnapshot();
  });
});
