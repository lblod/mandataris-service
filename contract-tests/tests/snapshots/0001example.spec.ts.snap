// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`fractions get mandataris fraction 1`] = `"{"fracties":["673C4FDF9AA6EE1DF1856B4E"]}"`;

exports[`fractions update mandataris publication status 1`] = `"{"status":"ok"}"`;

exports[`fractions update mandataris publication status 2`] = `
[
  {
    "g": {
      "type": "uri",
      "value": "http://mu.semte.ch/graphs/organizations/5116efa8-e96e-46a2-aba6-c077e9056a96/LoketLB-mandaatGebruiker",
    },
    "mandataris": {
      "type": "uri",
      "value": "http://data.lblod.info/id/mandatarissen/231e42a4-25a6-424b-9885-c16bea74a545",
    },
    "someone": {
      "type": "uri",
      "value": "http://data.lblod.info/id/bestuurseenheden/5116efa8-e96e-46a2-aba6-c077e9056a96",
    },
  },
  {
    "g": {
      "type": "uri",
      "value": "http://mu.semte.ch/graphs/organizations/5116efa8-e96e-46a2-aba6-c077e9056a96/LoketLB-mandaatGebruiker",
    },
    "mandataris": {
      "type": "uri",
      "value": "http://data.lblod.info/id/mandatarissen/2f40df94-a8c6-47bc-8771-493c6b5f29bc",
    },
    "someone": {
      "type": "uri",
      "value": "http://data.lblod.info/id/bestuurseenheden/5116efa8-e96e-46a2-aba6-c077e9056a96",
    },
  },
]
`;

exports[`fractions update mandataris publication status 3`] = `
[
  [
    {
      "deletes": [
        {
          "graph": {
            "type": "uri",
            "value": "http://mu.semte.ch/graphs/organizations/5116efa8-e96e-46a2-aba6-c077e9056a96/LoketLB-mandaatGebruiker",
          },
          "object": {
            "type": "uri",
            "value": "http://data.lblod.info/id/concept/MandatarisPublicationStatusCode/d3b12468-3720-4cb0-95b4-6aa2996ab188",
          },
          "predicate": {
            "type": "uri",
            "value": "http://lblod.data.gift/vocabularies/lmb/hasPublicationStatus",
          },
          "subject": {
            "type": "uri",
            "value": "http://data.lblod.info/id/mandatarissen/231e42a4-25a6-424b-9885-c16bea74a545",
          },
        },
        {
          "graph": {
            "type": "uri",
            "value": "http://mu.semte.ch/graphs/organizations/5116efa8-e96e-46a2-aba6-c077e9056a96/LoketLB-mandaatGebruiker",
          },
          "object": {
            "type": "uri",
            "value": "http://data.lblod.info/id/concept/MandatarisPublicationStatusCode/d3b12468-3720-4cb0-95b4-6aa2996ab188",
          },
          "predicate": {
            "type": "uri",
            "value": "http://lblod.data.gift/vocabularies/lmb/hasPublicationStatus",
          },
          "subject": {
            "type": "uri",
            "value": "http://data.lblod.info/id/mandatarissen/2f40df94-a8c6-47bc-8771-493c6b5f29bc",
          },
        },
      ],
      "inserts": [
        {
          "graph": {
            "type": "uri",
            "value": "http://mu.semte.ch/graphs/organizations/5116efa8-e96e-46a2-aba6-c077e9056a96/LoketLB-mandaatGebruiker",
          },
          "object": {
            "type": "literal",
            "value": "http://data.lblod.info/id/besluit/1",
          },
          "predicate": {
            "type": "uri",
            "value": "http://lblod.data.gift/vocabularies/lmb/linkToBesluit",
          },
          "subject": {
            "type": "uri",
            "value": "http://data.lblod.info/id/mandatarissen/231e42a4-25a6-424b-9885-c16bea74a545",
          },
        },
        {
          "graph": {
            "type": "uri",
            "value": "http://mu.semte.ch/graphs/organizations/5116efa8-e96e-46a2-aba6-c077e9056a96/LoketLB-mandaatGebruiker",
          },
          "object": {
            "type": "uri",
            "value": "http://data.lblod.info/id/concept/MandatarisPublicationStatusCode/9d8fd14d-95d0-4f5e-b3a5-a56a126227b6",
          },
          "predicate": {
            "type": "uri",
            "value": "http://lblod.data.gift/vocabularies/lmb/hasPublicationStatus",
          },
          "subject": {
            "type": "uri",
            "value": "http://data.lblod.info/id/mandatarissen/231e42a4-25a6-424b-9885-c16bea74a545",
          },
        },
        {
          "graph": {
            "type": "uri",
            "value": "http://mu.semte.ch/graphs/organizations/5116efa8-e96e-46a2-aba6-c077e9056a96/LoketLB-mandaatGebruiker",
          },
          "object": {
            "type": "literal",
            "value": "http://data.lblod.info/id/besluit/1",
          },
          "predicate": {
            "type": "uri",
            "value": "http://lblod.data.gift/vocabularies/lmb/linkToBesluit",
          },
          "subject": {
            "type": "uri",
            "value": "http://data.lblod.info/id/mandatarissen/2f40df94-a8c6-47bc-8771-493c6b5f29bc",
          },
        },
        {
          "graph": {
            "type": "uri",
            "value": "http://mu.semte.ch/graphs/organizations/5116efa8-e96e-46a2-aba6-c077e9056a96/LoketLB-mandaatGebruiker",
          },
          "object": {
            "type": "uri",
            "value": "http://data.lblod.info/id/concept/MandatarisPublicationStatusCode/9d8fd14d-95d0-4f5e-b3a5-a56a126227b6",
          },
          "predicate": {
            "type": "uri",
            "value": "http://lblod.data.gift/vocabularies/lmb/hasPublicationStatus",
          },
          "subject": {
            "type": "uri",
            "value": "http://data.lblod.info/id/mandatarissen/2f40df94-a8c6-47bc-8771-493c6b5f29bc",
          },
        },
      ],
    },
  ],
]
`;

exports[`fractions update-mandataris-fractie 1`] = `""`;

exports[`fractions update-mandataris-fractie 2`] = `
[
  [
    {
      "deletes": [],
      "inserts": [
        {
          "graph": {
            "type": "uri",
            "value": "http://mu.semte.ch/graphs/organizations/5116efa8-e96e-46a2-aba6-c077e9056a96/LoketLB-mandaatGebruiker",
          },
          "object": {
            "type": "uri",
            "value": "http://data.lblod.info/id/fracties/7e82f4b4-8424-4fbf-9cf4-dad695b53fbb",
          },
          "predicate": {
            "type": "uri",
            "value": "http://mu.semte.ch/vocabularies/ext/lmb/currentFracties",
          },
          "subject": {
            "type": "uri",
            "value": "http://data.lblod.info/id/personen/f5979b66f11984b5d21da224a3ce9fddda42e0c883fa1cdab8e1ff6c19179e3a",
          },
        },
      ],
    },
  ],
]
`;
